lib/timetable-form/Form.tsx

```diff
@@ -1,36 +1,58 @@
-import { useState } from "react";
-import "govuk-frontend/dist/govuk/govuk-frontend.min.css";
+import { useCallback, useMemo, useState } from "react";
 
 import { Button } from "../gds-components/button/Button";
 import { TextInput } from "../gds-components/text-input/TextInput";
 import { DateInput } from "../gds-components/date-input/DateInput";
-import { Stages, FormData } from "../types/timetable";
 import {
-  formStateToDevelopmentPlanTimetables,
-  devPlanToCSVString,
-} from "../utils/timetable";
-import { stageNames } from "../constants";
+  DevelopmentPlan,
+  DevelopmentPlanTimetable,
+  StageName,
+} from "../types/timetable";
+import { devPlanToCSVString } from "../utils/timetable";
+import { DEFAULT_DEVELOPMENT_PLAN } from "../constants";
+import { PlanViewer } from "../timetable-visualisation/PlanViewer";
 
 import styles from "./styles.module.css";
+import "govuk-frontend/dist/govuk/govuk-frontend.min.css";
 
-const defaultDate = "";
-
-const initialStages = Object.fromEntries(
-  stageNames.map((stage) => [stage, defaultDate])
-) as Stages;
+const {
+  timetableEvents: timetableEventsInitialState,
+  ...developmentPlanInitialState
+} = DEFAULT_DEVELOPMENT_PLAN;
 
 export const Form = (props: React.HTMLAttributes<HTMLDivElement>) => {
   const { className, ...otherProps } = props;
-  const [lpa, setLpa] = useState<string>("");
-  const [stages, setStages] = useState<Stages>(initialStages);
 
-  const getTimetableDownload = (formData: FormData) => {
-    const timetables = formStateToDevelopmentPlanTimetables(formData);
+  const [developmentPlan, setDevelopmentPlan] = useState<
+    Omit<DevelopmentPlan, "timetableEvents">
+  >(developmentPlanInitialState);
+  const [developmentPlanEvents, setDevelopmentPlanEvents] = useState<
+    DevelopmentPlanTimetable[]
+  >(timetableEventsInitialState);
 
-    const timetableCSV = devPlanToCSVString(timetables);
+  const timetableDownloadLink = useMemo(() => {
+    const timetableCSV = devPlanToCSVString({
+      ...developmentPlan,
+      timetableEvents: developmentPlanEvents,
+    });
 
     return `data:text/csv;charset=urf-8, ${timetableCSV}`;
-  };
+  }, [developmentPlan, developmentPlanEvents]);
+
+  const updateEventDate = useCallback(
+    (dateValue: string, stageName: StageName) => {
+      const updatedEvents = [...developmentPlanEvents];
+      const dateToUpdate = updatedEvents.find(
+        (event) => event.developmentPlanEvent === stageName
+      );
+
+      if (dateToUpdate) {
+        dateToUpdate.eventDate = dateValue;
+        setDevelopmentPlanEvents(updatedEvents);
+      }
+    },
+    [developmentPlanEvents]
+  );
```

Comment:
This is still a mutation, spreading on line 44 is only spreading the top level array but the references for each item will be the same. I think we'd need something like:

```suggestion
  const indexToUpdate = developmentPlanEvents.findIndex(
        (event) => event.developmentPlanEvent === stageName
      );
      if (indexToUpdate < 0) {
        return;
      }

      const updatedDate = {
        ...developmentPlanEvents[indexToUpdate],
        eventDate: dateValue,
      };
      const updatedEvents = [
        ...developmentPlanEvents.slice(0, indexToUpdate),
        updatedDate,
        ...developmentPlanEvents.slice(indexToUpdate + 1),
      ];
      setDevelopmentPlanEvents(updatedEvents);
```

But I'm now starting to wonder if what you had before was neater, even if it means some extra looping, what do you think?

Reply:
the .map approach was definitely neater, I agree, and I am not a big fan of the slicing using the index approach

Reply:
another argument for the "slicing" approach is that we get to react if indexToUpdate is not found for some reason or eventually if more than one is found

Reply:
I think we might have to think separate about updating (I'm assuming that's what you're referring to with more than one being found), niether approach we've used so far would be appropriate.

For now, I think our options are stick with the map approach you had or go with something like the suggestion above, I'm happy with either so I'll leave it up to you :)

Reply:
I changed it back to use map

Reply:
Great, it's likely we'll have to take another look at this later when we implement editing